sycl/include/CL/sycl/detail/buffer_usage.hpp

```diff
@@ -0,0 +1,75 @@
+//==------------ buffer_usage.hpp - SYCL standard header file ------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/context.hpp>
+
+#include <deque>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+class handler;
+namespace detail {
+
+// allows us to represent whether a user has set (or not) a boolean value, and,
+// if so, to what.
+enum class settable_bool { set_false = -1, not_set, set_true };
+
+// acts as return value of whenCopyBack(buff*).  which returns dtor, immediate
+// or never acts as the 'now' parameter for copyBackSubBuffer(now, buff*),
+//  because that function will be called from two places (dtor and immedaite)
+// acts as the 'when' parameter for shouldCopyBack(when, buff*) (dtor, immediate
+// and undertermined)
+//  where 'undetermined' is simply "will this buffer copy back at any time?"
+enum class when_copyback {
+  dtor,      // addCopyBack during sub-buffer dtor
+  immediate, // copy-back achieved by enqueued map operation
+  never
+};
+
+// need to track information about a sub/buffer,
+// even after its destruction, we may need to know about it.
```

Comment:
The sub-buffers handle the copy-back of their own data.  They will likely have been destroyed before the buffer_impl is destroyed.  And when it is destroyed it needs to know if it has to initiate a copy back or not.  So before it can answer that question it needs to know if there were any sub-buffers and if they actually performed any copy operations (ie had any write accessors declared).